src/requests/salerequest.cpp

```diff
@@ -0,0 +1,71 @@
+#include "salerequest.h"
+#include "requestdefines.h"
+#include "requesttools.h"
+
+namespace graft {
+
+GRAFT_DEFINE_IO_STRUCT(SaleRequest,
+    (std::string, Address),
+    (std::string, SaleDetails),
+    (std::string, Amount)
+);
+
+GRAFT_DEFINE_IO_STRUCT(SaleResponse,
+    (uint64, BlockNumber), //TODO: Need to check if we really need it.
+    (std::string, PaymentID)
+);
+
+Router::Status workerHandler(const Router::vars_t& vars, const graft::Input& input,
+                            graft::Context& ctx, graft::Output& output)
+{
+    SaleRequest in = input.get<SaleRequest>();
+    if (!in.Address.empty() && !in.Amount.empty())
+    {
+        std::string payment_id = generatePaymentID();
+        if (!ctx.global.hasKey(payment_id + CONTEXT_KEY_SALE))
+        {
+            ErrorResponse err;
+            err.code = ERROR_SALE_REQUEST_FAILED;
+            err.message = ERROR_MESSAGE(MESSAGE_SALE_REQUEST_FAILED);
+            output.load(err);
+            return Router::Status::Error;
+        }
+        uint64_t amount = convertAmount(in.Amount);
+        if (amount <= 0)
+        {
+            ErrorResponse err;
+            err.code = ERROR_AMOUNT_INVALID;
+            err.message = ERROR_MESSAGE(MESSAGE_AMOUNT_INVALID);
+            output.load(err);
+            return Router::Status::Error;
+        }
+        SaleData data(in.Address, 0, amount);
+        ctx.global[payment_id + CONTEXT_KEY_SALE] = data;
+        ctx.global[payment_id + CONTEXT_KEY_SALE_STATUS] = RTAStatus::InProgress;
+        if (!in.SaleDetails.empty())
+        {
+            ctx.global[payment_id + CONTEXT_KEY_SALE_DETAILS] = in.SaleDetails;
+        }
+        SaleResponse out;
+        out.BlockNumber = data.BlockNumber;
+        out.PaymentID = payment_id;
+        output.load(out);
+        return Router::Status::Ok;
+    }
+    else
+    {
+        ErrorResponse err;
+        err.code = ERROR_INVALID_PARAMS;
+        err.message = ERROR_MESSAGE(MESSAGE_INVALID_PARAMS);
+        output.load(err);
+        return Router::Status::Error;
+    }
+}
+
+void registerSaleRequest(graft::Router &router)
+{
+    Router::Handler3 *h3_test = new Router::Handler3(nullptr, workerHandler, nullptr);
+    router.addRoute("/dapi/sale", METHOD_POST, h3_test);
+}
```

Comment:
Looks like new without delete. Also, is it assumed that registerSaleRequest is called once, or multiple calls are not a harm? Please, consider something like this:
static done = false; if(done) return;
done =true;
static Router::Handler3 h3_test ( Router::Handler3(nullptr, workerHandler, nullptr) );
router.addRoute(... , &h3_test);

Otherwise it will be difficult to find real and such memory leaks.

Reply:
It can be called more than one time.

Also Router::Handler3 managed by Router